wasmi-impl/src/lib.rs

```diff
@@ -17,28 +28,62 @@ impl From<wasmi::Error> for ExecWasmError {
     }
 }
 
-pub fn exec_wasm_with_data(
+struct Proc {
+    out_buffer: Vec<u8>,
+    return_code: Option<RuntimeValue>,
+}
+
+impl<'a> Proc {
+    pub fn bytes(&'a self) -> impl 'a + Iterator {
+        self.out_buffer.iter()
+    }
+}
+
+pub fn exec_wasm_with_data<'a>(
     binary: &[u8],
     data: &[u8],
-) -> Result<Option<RuntimeValue>, ExecWasmError> {
+    out_size: usize, /* output size in bytes */
+) -> Result<Proc, ExecWasmError> {
     let module = wasmi::Module::from_buffer(binary)?;
 
-    // TODO: Calculate the memory size always be larger than `data`
-    let mem_instance = MemoryInstance::alloc(Pages(200), None)?;
+    let data_size = data.len();
+
+    /*
+     *  Calculate the memory size to always be larger than `data`.  Small
+     *  values (a WASM numeric vector or smaller) should always fit.
+     */
+    let calc_rem = |num_bytes| cmp::max(num_bytes, MEM_SMALL_BYTES) % PAGE_SIZE_BYTES;
+    let calc_pages = |num_bytes| cmp::max(num_bytes, MEM_SMALL_BYTES) / PAGE_SIZE_BYTES;
+    let combined_size = data_size + out_size;
+    let num_pages = cmp::min(
+        MAX_PAGES,
+        match calc_rem(combined_size) {
+            0 => calc_pages(combined_size),
+            _ => calc_pages(combined_size) + 1,
+        },
+    );
```

Comment:
Could we simplify it to something like this, with appropriate casts?

```suggestion
    let combined_size = data_size + out_size;
    let num_pages = (combined_size as f64 / PAGE_SIZE_BYTES as f64).ceil() as usize;
```

We can add a TODO to use [`int_roundings`](https://github.com/rust-lang/rust/issues/88581) once that's stabilized.

Reply:
We should also probably have it panic rather saturate to `MAX_PAGES`?

Reply:
(Edited to use checked addition.)